font/src/directwrite/mod.rs

```diff
@@ -107,73 +242,32 @@ impl crate::Rasterize for DirectWriteRasterizer {
 
                     idx += 1;
                 }
-            },
+            }
         }?;
 
-        let face = font.create_font_face();
-        self.fonts.push(face);
+        let font_face = font.create_font_face();
+        let family_name = font.family_name();
+        let weight = font.weight();
+        let style = font.style();
+        let stretch = font.stretch();
+        let key = FontKey::next();
+        self.keys.insert(desc.clone(), key);
+        self.fonts.insert(key, LoadedFont { font_face, family_name, weight, style, stretch });
```

Comment:
All these variables are defined only to be used in the `LoadedFont` initialization. So they shouldn't be defined as separate variables but we should just inline these in the struct definition.

However since all we seem to be doing is querying fields on one thing to create another, this seems like a prime candidate for an `Into` implementation?

I'd assume there's a reason why we can't just use dwrote's `Font` ourselves, right? Otherwise that would be the obvious solution.

Reply:
I am caching all of these things because otherwise it creates a lot of additional directwrite API calls to retrieve this information every time a fallback font lookup is needed (all this information is input to the `map_characters` API call).

This was done in the name of performance.

In addition I think `font.create_font_face()` is likely to be relatively heavyweight so would prefer avoid doing that on every glyph rasterization, which we'd have to do if we just store `dwrote::Font`.

I really like the `Into` idea though.

Reply:
That's fair enough, I'm not aware of how many of these methods are basically just a memory access versus an actual computation. `Into` would definitely be a good solution either way.

Reply:
A lot of the `dwrote` methods involving `String` have at least one additional conversion going on to convert back and forth from UTF-16 C++ encoding to Rust `String`. Based on experiments I've done, caching all these saves us a number of microseconds, which counts for a lot when the comment above was asking me to try to make perf ~ 20-30 us per fallback lookup!

Will go down the `Into` road.

Reply:
Great job actually looking into this, I'll definitely gladly take those free performance improvements.